### PR TITLE
Prefer local wildcard audits to remote (imported) audits.

### LINKED ISSUE #587: prefer local wildcard audits to remote regular audits
Over on [mozilla-central](https://phabricator.services.mozilla.com/D197955), @jrmuizel and @glandium identified an interesting issue where Jeff was trying to replace a regular audit with a wildcard audit. Rather than reifying the local wildcard audit, cargo-vet instead re-fetches the locally-deleted regular audit from the [aggregated mozilla-wide set](https://raw.githubusercontent.com/mozilla/supply-chain/main/audits.toml) and re-inserts it as an import.

I'm assuming this is because the precedence order is currently:
1. local regular audits
2. remote regular audits
3. local wildcard audits
4. remote wildcard audits

Ideally we'd swap 2 and 3.

@afranchuk Is my hypothesis correct?

### PR DESCRIPTION
**Pull request by afranchuk**
Closes #587.

### CODE REVIEW

File: src/tests/import.rs [hunk 1]
```diff
@@ -1934,3 +1934,76 @@ fn import_criteria_map_aggregated_error() {
...
+            "third-party2".to_owned(),
+            vec![wildcard_audit(1, SAFE_TO_DEPLOY)],
+        )]
+        .into(),
+        trusted: SortedMap::new(),
+    };
+
+    imports
+        .audits
+        .insert(FOREIGN.to_owned(), foreign_audits.clone());
```
Comment (mystor) [inline src/tests/import.rs:1973]: To mode accurately model the situation which originally brought this up, we don't need any audits or their corresponding publishers pre-populated within `imports.lock`. We also don't need the foreign wildcard entry to be present.

File: src/tests/import.rs [hunk 2]
```diff
@@ -1934,3 +1934,76 @@ fn import_criteria_map_aggregated_error() {
...
+
+    imports.publisher.insert(
+        "third-party2".to_owned(),
+        vec![publisher_entry_named(
+            ver(DEFAULT_VER),
+            1,
+            "user1",
+            "User One",
+        )],
+    );
```
Comment (mystor) [inline src/tests/import.rs:1983]: This publisher entry shouldn't be pre-populated within `imports.lock` - we should be fetching it from the network to model the situation better.

File: src/tests/import.rs [hunk 3]
```diff
@@ -1934,3 +1934,76 @@ fn import_criteria_map_aggregated_error() {
...
+        audits: [(
+            "third-party2".to_owned(),
+            vec![full_audit(ver(DEFAULT_VER), SAFE_TO_DEPLOY)],
+        )]
+        .into(),
+        wildcard_audits: [(
+            "third-party2".to_owned(),
+            vec![wildcard_audit(1, SAFE_TO_DEPLOY)],
+        )]
+        .into(),
```
Comment (mystor) [inline src/tests/import.rs:1967]: I'm pretty sure this foreign wildcard audit is unnecessary, but I suppose it doesn't hurt to include it in the test.

File: src/tests/import.rs [hunk 4]
```diff
@@ -1934,3 +1934,76 @@ fn import_criteria_map_aggregated_error() {
...
+        .package(
+            "third-party2",
+            &[reg_published_by(ver(DEFAULT_VER), Some(1), "2022-12-15")],
+        )
+        .serve(&mut network);
+    network.mock_serve_toml(FOREIGN_URL, &foreign_audits);
+
+    let store = Store::mock_online(&cfg, config, audits, imports, &network, true).unwrap();
+
+    let output = get_imports_file_changes_prune(&metadata, &store);
```
Comment (mystor) [inline src/tests/import.rs:2007]: Theoretically you shouldn't need to do a prune for the correct outcome to happen here.

File: src/resolver.rs [hunk 5]
```diff
@@ -1446,6 +1447,7 @@ fn search_for_path(
     #[derive(Debug, Copy, Clone, Eq, PartialEq, Ord, PartialOrd)]
     enum CaveatLevel {
         None,
+        Import,
```
Comment (mystor) [inline src/resolver.rs:1450]: I think this is ranked too high. A new import should never be ranked higher than something like a non-importable audit or preferred exemption/unpublished, as that means that unnecessary change to the imports.lock would happen without doing a prune.

**Review by mystor — CHANGES_REQUESTED** [review]

File: src/tests/import.rs (hunk 2 above)
Comment (afranchuk) [reply src/tests/import.rs:1983]: Oh right, so I was testing in a situation where we did actually already have it there (import first, then add a wildcard audit). I can change this or make a separate test to test fresh imports.

File: src/tests/import.rs (hunk 1 above)
Comment (afranchuk) [reply src/tests/import.rs:1973]: The foreign wildcard entry was kind of to "cover all bases", though somewhat lazily.

File: src/tests/import.rs (hunk 4 above)
Comment (afranchuk) [reply src/tests/import.rs:2007]: We do need to prune for the correct outcome when the import is not fresh.

File: src/resolver.rs (hunk 5 above)
Comment (afranchuk) [reply src/resolver.rs:1450]: A fresh import is marked `FreshImport`, this is for existing imports. They previously had `CaveatLevel::None`, so I thought this was appropriate.

**Comment by afranchuk** [comment]
I've added a second test. Now one tests against existing imports and the other tests against a fresh import.

File: src/resolver.rs [hunk 6]
```diff
@@ -1177,7 +1177,8 @@ impl<'a> AuditGraph<'a> {
                         audit_index,
                         publisher_index,
                     };
-                    let is_fresh_import = entry.is_fresh_import || publisher.is_fresh_import;
+                    let is_fresh_import = import_index.is_some()
+                        && (entry.is_fresh_import || publisher.is_fresh_import);
```
Comment (mystor) [inline src/resolver.rs:1181]: (I meant to include a comment like this when I first submitted the review, but apparently it didn't save - apologies)

We shouldn't consider a local wildcard import which has a fresh import to be a non-fresh import. Something like that requires changes to `imports.lock` in order to use, so should be considered fresh.

Perhaps we could change `DeltaEdge::is_fresh_import` to instead be an enum like:

```rust
#[derive(Debug, Copy, Clone, Eq, PartialEq)]
enum DeltaEdgeFreshness {
  // All information requried for this delta edge is already stored within the store's supply-chain.
  // Using this edge will require no changes to imports.lock.
  Stale,
  // This edge originates from an already-imported or local wildcard audit,
  // however will require importing fresh publisher information into imports.lock to use.
  FreshPublisher,
  // This edge is fully fresh, and will require adding the audit entry to imports.lock to use.
  Fresh,
}

impl DeltaEdgeFreshness {
  fn is_fresh(&self) -> bool { self != DeltaEdgeFreshness::Stale }
}
```

For most of the logic, checks of `is_fresh_import` can be replaced with the equivalent of `is_fresh()` (or just inlining that check), but the caveat level match pattern would need to be changed, like:

```rust
                _ => match edge.is_fresh_import {
                   DeltaEdgeFreshness::Stale => CaveatLevel::None,
                   DeltaEdgeFreshness::FreshPublisher => CaveatLevel::FreshPublisher,
                   DeltaEdgeFreshness::Fresh => CaveatLevel::FreshImport,
               }
```

File: src/resolver.rs (hunk 5 above)
Comment (mystor) [reply src/resolver.rs:1450]: I've added a bit more elaboration as to what I was going for in my comment above around this being too high of a ranking (tl;dr with the current implementation, this isn't just for existing imports, but also for local wildcard audits with fresh publisher entries, which definitely shouldn't be above these other cases).

I don't think we need a ranking for fully stale imports other than `None`, though I'm open to an argument for where that would be useful.

**Review by mystor — CHANGES_REQUESTED** [review]

File: src/resolver.rs (hunk 6 above)
Comment (afranchuk) [reply src/resolver.rs:1181]: This is the change I was most hesitant about, so I'm glad you had some opinions on it.

File: src/resolver.rs (hunk 5 above)
Comment (afranchuk) [reply src/resolver.rs:1450]: If we don't rank fully stale imports other than `None`, a local wildcard audit won't be preferred over an existing import. So if you already have some imports set up but then add a local wildcard audit, the imports will remain and the local wildcard audit would be unused (and prunable). I think we want the local wildcard audit to be preferred.

File: src/resolver.rs [hunk 7]
```diff
@@ -341,6 +341,40 @@ pub enum RequiredEntry {
...
+        }
+    }
+
+    pub fn fresh_publisher(is_fresh_publisher: bool) -> Self {
+        if is_fresh_publisher {
+            DeltaEdgeFreshness::FreshPublisher
+        } else {
+            DeltaEdgeFreshness::Stale
+        }
+    }
```
Comment (mystor) [inline src/resolver.rs:371]: Perhaps this one should handle both sides, rather than just the publisher one?

Also not sure why these methods are public given the type they are on is not exposed outside of this module, so let's drop the `pub` annotations.

```suggestion
   fn fresh_publisher(is_fresh_audit: bool, is_fresh_publisher: bool) -> Self {
        if is_fresh_audit {
            DeltaEdgeFreshness::Fresh
        } else if is_fresh_publisher {
            DeltaEdgeFreshness::FreshPublisher
        } else {
            DeltaEdgeFreshness::Stale
        }
    }
```

File: src/resolver.rs [hunk 8]
```diff
@@ -1203,18 +1242,19 @@ impl<'a> AuditGraph<'a> {
                     let to_ver = Some(&publisher.version);
                     let criteria = criteria_mapper.criteria_from_list(&entry.criteria);
                     let origin = DeltaEdgeOrigin::Trusted { publisher_index };
+                    let freshness = DeltaEdgeFreshness::fresh_publisher(publisher.is_fresh_import);
```
Comment (mystor) [inline src/resolver.rs:1245]: I'm not sure we want to use `fresh_publisher` here. Technically it is somewhat "correct", but a trusted entry is less validated than a normal audit, so we might not want to be prioritizing them in this way?

I believe that if we put them at the same caveat level as other audits, it'll be picked below normal audits due to the tie-breaker ordering by `DeltaEdgeOrigin`, which is what we want:

https://github.com/mozilla/cargo-vet/blob/67e92b6d03f48802907b2c80119ed328289a114d/src/resolver.rs#L301-L302

https://github.com/mozilla/cargo-vet/blob/67e92b6d03f48802907b2c80119ed328289a114d/src/resolver.rs#L1486

```suggestion
                    let freshness = DeltaEdgeFreshness::fresh(publisher.is_fresh_import);
```

**Review by mystor — CHANGES_REQUESTED** [review]

**Comment by afranchuk** [comment]
@mystor does this look better now?

**Review by mystor — APPROVED** [review]
